Update browser cache on backend Permission & CodeType invalidate

mvilliger opened the pull request:
373498

mvilliger committed:
Update browser cache on CodeType and Permission invalidate

373498

cguglielmo commented on eclipse-scout-core/src/code/codes.ts:
```diff
@@ -41,14 +43,23 @@ export const codes = {
       };
     }
-    codes.init(data);
+    codes.init(data)
```
Subscribe should be done before codes are initialized. Otherwise there is a chance that the ui notification is missed if the codes change between init and subscribe

cguglielmo commented on eclipse-scout-core/src/code/codes.ts:
```diff
@@ -25,12 +25,14 @@ export const codes = {
 
   /** @internal */
-  _preInit(url: string, data: any) {
+  _handleBootstrapResponse(url: string, data: any) {
```
the _preInit name is used for texts.ts and locales.ts as well. Please change these to _handleBootstrapResponse as well

cguglielmo commented on org.eclipse.scout.rt.uinotification/src/main/java/org/eclipse/scout/rt/api/uinotification/UiNotificationRegistry.java:
```diff
@@ -372,6 +377,21 @@ public void removeListener(String topic, UiNotificationListener listener) {
   }
 
+  /**
+   * Retrieves how many observer are listening for the topic given.
```
observer -> observers
topic given -> given topic
 for given topic. -> for the given topic.

cguglielmo commented on org.eclipse.scout.rt.uinotification/src/main/java/org/eclipse/scout/rt/api/uinotification/UiNotificationRegistry.java:
```diff
@@ -185,14 +188,14 @@ protected List<UiNotificationDo> get(String topic, String user, final List<UiNot
+   * that needs to be sent as soon as the connection is reestablished again.
    * <p>
-   * The result can contain multiple notifications, one for each cluster node.
-   * So, if a cluster node created notifications for the topic, the last one for that node will be returned.
```
Why have you adjusted this? I find it more readable, if every sentence is on one line. We do the same for markdown and ascii doc text.

cguglielmo commented on org.eclipse.scout.rt.uinotification/src/test/java/org/eclipse/scout/rt/api/uinotification/UiNotificationRegistryTest.java:
```diff
@@ -49,6 +49,23 @@ public void before() {
   }
 
+  @Test
+  public void testIsListenedForTopic() {
```
rename to testGetListenerCount

cguglielmo commented on eclipse-scout-core/src/security/AccessControl.ts:
```diff
@@ -63,10 +67,41 @@ export class AccessControl extends PropertyEventEmitter implements AccessControl
 
   startSync() {
     this._sync();
+    this._subscribeForNotifications();
```
subscribe first, then sync

cguglielmo commented on eclipse-scout-core/src/security/AccessControl.ts:
```diff
@@ -25,16 +25,18 @@ export class AccessControl extends PropertyEventEmitter implements AccessControl
     super();
     this.permissionsUrl = null;
-    this.interval = 1800000; // 30 * 60 * 1000 (30 minutes)
+    this.interval = 1_800_000; // 30 * 60 * 1000 (30 minutes)
```
I think we should remove this.interval, because it does not reload every 30 min anymore. This variable now only specifies the limit for the retryIntervals array.
Actually, I don't think it should ever stop retrying, unless the session is invalidated.

cguglielmo commented on eclipse-scout-core/src/security/AccessControl.ts:
```diff
@@ -52,9 +54,11 @@ export class AccessControl extends PropertyEventEmitter implements AccessControl
+      let limit = interval / factor; // retryInterval is the time between retries
+      while (retryInterval < limit) {
         retryIntervals.push(retryInterval);
-        retryInterval = retryInterval * 2;
```
Doubling the retryInterval is probably not ideal, because it will soon be a very large interval. I know, it is nothing new, but I think we should change it. Maybe we have to adjust Call.ts to support this requirement, e.g. we could set maxRetries to -1 which means no limit and update this member based on the retryInterval only if maxRetries is null

cguglielmo commented on eclipse-scout-core/src/security/AccessControl.ts:
```diff
@@ -63,10 +67,41 @@ export class AccessControl extends PropertyEventEmitter implements AccessControl
+  }
+
+  protected _onPermissionUpdateNotify(event: UiNotificationEvent) {
+    let reloadDelayWindow = scout.nvl(event.message['reloadDelayWindow'], 0) as number;
```
Better create PermissionUpdateNotificationDo.ts which contains reloadDelayWindow as typed member to make clear what notification is expected

cguglielmo commented on eclipse-scout-core/src/uinotification/UiNotificationPoller.ts:
```diff
@@ -109,6 +109,7 @@ export class UiNotificationPoller extends PropertyEventEmitter {
   }
 
   protected _poll() {
+    this._call?.abort(); // abort in case there is already a call running
```
This should be fine but what makes it necessary? How can it happen?

cguglielmo commented on eclipse-scout-core/src/code/codes.ts:
```diff
@@ -84,16 +60,17 @@ export const codes = {
+  /**
+   * @returns the global {@link CodeTypeCache} instance. If required, a new one is created (on first use).
+   */
+  getCodeTypeCache(): CodeTypeCache {
```
I would just name it getCache.

cguglielmo commented on eclipse-scout-core/src/code/CodeTypeCache.ts:
```diff
@@ -0,0 +1,119 @@
+
+  constructor() {
+    super();
+    this.events.registerSubTypePredicate('codeTypeAdd', (event: CodeTypeAddEvent, codeTypeId) => event?.codeType?.id === codeTypeId); // only works if the CodeTypeId can be converted to a string
```
How can event and codetype be null? I don't see the reason why it is necessary to add the ? operator

cguglielmo commented on eclipse-scout-core/src/util/Call.ts:
```diff
@@ -64,15 +66,13 @@ export abstract class Call implements CallModel, ObjectWithType {
       this.maxRetries = this.retryIntervals.length;
     }
+    if (objects.isNullOrUndefined(this.defaultRetryInterval)) {
+      this.defaultRetryInterval = 300;
```
Just set it in constructor?

mvilliger replied:
The idea behind this is, that if the model contains null for defaultRetryInterval, the default don't get overwritten with null. What do you mean?

mvilliger committed:
Expose ConfigProperties to Scout JS

- Add new config utility in Scout JS to access config properties from
  backend systems
- Add new systems utility to configure backend API urls
- Add REST resource and ResourceLoader to fetch exposed config
  properties from UI server and backend server
- Add contributor to expose config properties
- Add function to convert IConfigProperties to ConfigPropertyDo
- Replace existing properties sent to the browser with new config
  utility:
  - devMode
  - ui.backgroundPollingMaxWaitTime
  - uinotification.waitTimeout

373066
